File: Daily_Coding_Problem-05.cpp
/* MARK: - SOLUTION ✨ */

#include "Daily_Coding_Problem.h"

// MARK: - Solution

// This is the implementation of cons function
pair<int, int>cons(int a, int b) {
    return make_pair(a, b);
}

// This is the implementation of car function that returns the first element of the pair (a)
int car(pair<int, int> p) {
    return p.first;
}

// This is the implementation of cdr function that returns the last element of the pair (b)
int cdr(pair<int, int> p) {
    return p.second;
}

// MARK: - Main func
int main()
{
    pair<int, int> p = cons(3,4);
    cout << car(p) << endl;
    cout << cdr(p) << endl;
}

File: Daily_Coding_Problem-05.swift
#!/usr/bin/swift

/* MARK: - SOLUTION ✨ */

class Problem<T> {

    typealias Pair = ( ((T, T) -> T) -> T)  

    func cons(_ a: T, _ b: T) -> Pair {

        func pair(cb: (T, T) -> T) -> T {
            return cb(a, b)
        }
 
        return pair
    }

    /*func car(pair: Pair, firstElementHandler: @escaping (T) -> Void) { // retorna primeiro valor do par
        pair({ a, _ in
            firstElementHandler(a)
        })
    }*/

    func cdr(pair: Pair) -> T { // retorna ultimo valor do par
        return pair({ _, b in
            return b
        })
    }

}

let problem = Problem<Int>()
let pair = problem.cons(1,2)

/*
problem.car(pair: pair, firstElementHandler: {
    print($0)
})
*/

print(problem.cdr(pair: pair))